hibernate-core/src/main/antlr/hql-sql.g

```diff
@@ -436,7 +436,7 @@ selectExpr
 	;
 
 count
-    : #(COUNT  { inCount = true; } ( DISTINCT { inCountDistinct = true; } | ALL )? ( aggregateExpr | ROW_STAR ) ) {
+    : #(COUNT  { inCount = true; } ( DISTINCT { inCountDistinct = true; } | ALL )? ( aggregateExpr | ROW_STAR | query) ) {
```

Comment:
@beikov Right, so long as the query is correlated, this works at least for H2:
`SELECT COUNT( DISTINCT( SELECT b2.x FROM Bar2 b2 WHERE b2.barId = b.id ) ) FROM Bar b`

But again, portability to other databases is my concern, specifically when MSSQL documentation states: `Aggregate functions and subqueries are not permitted.` inside a `COUNT` function as a part of the expression BNF.

But I agree with you and @vladmihalcea that what the test outlines does not exactly correlate with the JIRA query presented however.